### PR TITLE
fix(db): repair stale board climb holds

### PR DESCRIPTION
**Pull request by marcodejongh**
## Summary

- repair stale and globally invalid `board_climb_holds` rows with a dry-run-first manifest, reviewed digest/count guards, writer-blocking transaction locks, and post-write verification
- project multi-frame Aurora climbs consistently across sync, Kilter Grips, unified imports, and backfills so the repaired rows do not drift again
- keep similar-climb matching and popular board counts correct while stale production rows await a separately approved repair run
- document the exact operator workflow, safeguards, and rollback boundary

Refs #3985

The issue stays open on purpose: this PR only lands the tooling and the guards. The production dry-run, the apply, the count verification, and the popular-config cache bust are separate operator steps that each need their own explicit approval, and #3985 tracks them until they are done.

## Release Notes

- Animated board climbs now keep the right holds for similar-climb matches and board counts after the guarded data repair is run.

## Test plan

- focused backend and active-writer suites: 134 tests passed
- focused DB helper/repair/import suites: 31 tests passed, including real PostgreSQL apply, verification, clean rerun, and rollback coverage
- full DB suite: 393 passed, 11 skipped, with the same pre-existing unrelated `create-climb-filters` assertion failures
- `vp test run --project kilter-sync`: 239 passed, 1 skipped
- `vp run typecheck` (whole monorepo)
- `vp check`: 0 errors
- `git diff --check`
- independent adversarial audit: no P0–P2 findings
- `vp run dev` loaded `.boardsesh/qa-notes.md`; homepage, backend health, and dev metadata returned 200

## Production gate

This PR does **not** run the production dry-run, apply the repair, refresh caches, or delete production data. Those operator actions remain a separate explicit approval after the dry-run digest and counts are reviewed.

### CODE REVIEW

**Comment by marcodejongh** [comment]
Addressed all four automated-review notes in `6641eaf30`:

- multi-frame empty projections now parse exactly once, with a resolver regression test
- the same-board legacy `frames = NULL` recovery boundary is explicit in the authoritative-frame contract
- locked-manifest drift testing matches the candidate SQL/state instead of a positional execute count and proves no mutation ran
- an unchanged authoritative empty projection now directly proves the row-derived SHA256(empty) fingerprint-clear branch

Focused validation passed: resolver 9/9, repair suites 21/21 (including real PostgreSQL apply/idempotence), scoped check with no errors, and `git diff --check`.

**Comment by marcodejongh** [comment]
Resolved the actionable second-pass feedback in `ddd884751` and clarified the intentional contracts:

- a top-level or Drizzle-wrapped PostgreSQL `55P03` now prints a retry-after-traffic-drops operator hint while preserving the original error and nonzero exit
- the similarity predicate is renamed to `isSupportedSimilarityHold`; it is intentionally narrower than popular-config/repair structural validity so unknown-but-well-formed future board states are not deleted or omitted from placement counts
- the snapshot comment now names the advisory-lock `SELECT` as the first snapshot-taking command; the table locks deliberately remain before it so already-in-flight writer commits are visible
- target frame parsing now has one shared mapping path after the multi-frame/materialized-row branch
- existing tests already cover single materialized, single legacy fallback, multi-frame direct, empty multi-frame exactly once, and invalid single rows

Validation: backend focused 30/30, repair focused 22/22 including real PostgreSQL apply/idempotence, backend typecheck, scoped check with no errors, and `git diff --check`. Independent triage found no P0/P1.

**Comment by marcodejongh** [comment]
Final review cleanup is in `7294f3764`:

- the nullable frame-count guard now reads in evaluation order
- frame-derived similarity holds apply the supported-state filter explicitly on both input paths
- empty persisted frames have direct sync coverage
- DB operator scripts now have a local TypeScript config, so staged checks resolve Node built-ins correctly

Focused sync/resolver validation passed 42/42, the staged `vp check --fix` hook passed, and `git diff --check` is clean. The repair behavior and production-data boundary are unchanged.

**Comment by marcodejongh** [comment]
Fable review pass — two code defects, one PR-body issue, one docs gap. All addressed in 488aef3.

**Critical — the strict projector would have refused the whole production apply.**
`strictlyProjectStoredRows` treated an empty frame 0 as malformed and compared `frames_count` against `parseFramesSegments().length`, which drops that leading empty slot. But an empty first frame is the delayed-start encoding this very PR's Grips writer emits: `decodeGripsClimbConcat('h10p13s2', …, 2)` produces `,"p100r13` with `framesCount: 2`. Every such climb raised two blockers (`frame 0 is an empty unquoted absolute frame` + `frames_count mismatch: stored=2 parsed=1`), and blockers have no override, so one delayed-start climb in production would have blocked the entire apply — on rows that were already canonical. Frame 0 may now be empty; an empty unquoted frame after frame 0 still blocks, and `frames_count` is compared against the raw comma-delimited slot count, which is what Aurora and Grips actually record. Regression tests: `strictlyProjectStoredRows('kilter', ',"p100r13', 2)` returns `[{holdId: 100, frameNumber: 0, holdState: 'HAND'}]`, the same manifest plans no mutation for it, and `p1r13,` still blocks.

**Major — `decideCatalogFingerprint` deduped on SHA256(empty).**
Now that `decodeGripsClimbConcat` returns projected holds, a climb whose projection is empty hashed to the constant SHA256(`''`) — the exact value the repair script documents as "not a valid climb identity" and clears to NULL. The first hold-less climb became its owner and every later one was silently written as an alias of an unrelated climb, re-introducing the fingerprint the repair had just cleared. Empty projections now skip dedup entirely: own canonical, `NULL` fingerprint, never entered into `fingerprintToCanonical` — the same rule `enrichFingerprintOwnersWithLegacyCompatibility` already enforced. `catalog-sync.ts` step 3 skips the `.set` and writes `holdFingerprint: null`. Two tests cover it: two distinct hold-less climbs stay independent, and a stranded stored SHA256(empty) owner is ignored rather than aliased onto.

**Major — `Closes #3985` → `Refs #3985`.**
The issue's deliverable is the production repair, which this PR deliberately does not run. Swapped, with a note that #3985 stays open to track the approved dry-run / apply / count-verify / cache-bust steps.

**Minor — runbook not in the docs index.**
Added `docs/board-climb-holds-repair.md` to CLAUDE.md's Documentation list, plus a paragraph in the runbook itself explaining that a leading empty frame is accepted and a later one still blocks.

**One extra, not from the review:** the new `packages/db/scripts/tsconfig.json` set `rootDir` to `packages/db`, which tripped TS6059 on three unrelated scripts that import sibling packages (`analyze-led-mapping.ts`, `migration-journal.ts`). PR CI diff-lints so it stayed green here, but the full-repo lint on a push to `main` would have caught it. `rootDir` now points at the monorepo root.

Validation: `vp check` 0 errors, `vp run typecheck` clean, `vp test run --project kilter-sync` 239 passed, `vp run test:db` 393 passed with only the pre-existing unrelated `create-climb-filters` failures.

**Comment by marcodejongh** [comment]
CI blocker is not this branch. `test-default` fails on `i18n catalog completeness > de > boards.json ships the same key set as en-US` — 30 keys missing from `packages/shared/i18n/locales/de/boards.json` (`mobile.gymPicker.*`, `mobile.create.duplicate.*`, `mobile.offline.*`, `boardForm.create.duplicate*`) plus 2 stale extras.

That gap lives on `main`: `de` joined `STRICTLY_ENFORCED_LOCALES` there while the German catalog stayed behind, and main's own CI run for 681a0481d is red on `test-default` and `lint` for the same reason. Since PR CI tests the merge commit, every open PR inherits it. #4213 already fixes it and is green — this PR goes green on a re-run once that lands. Nothing here touches i18n; every other check passes.

Staying in draft until then.

**Comment by marcodejongh** [comment]
🤖 Taking this over — bug-batch loop `moarbugs100500` (session 43ea5be5).

This PR has been sitting in flight, so this loop is adopting it: full re-review of the diff against current `main`, rebase and conflict resolution as needed, fixes for anything the review turns up, and shepherding through CI until it is ready to land.

Parallel automation (`moarbugs10000` and any other loop): please leave this PR to me — comment here if you were already on it and I will hand it back.

**Comment by marcodejongh** [comment]
Rebase-and-pnpm-migrate sweep: tried to rebase this PR onto post-migration `main`, hit a real conflict unrelated to the Bun→pnpm switch, and aborted rather than guess at a code-level merge. Branch is untouched, nothing was pushed.

Conflicting: packages/aurora-sync/src/sync/shared-sync.test.ts, packages/backend/src/graphql/resolvers/social/boards.ts, vite.config.ts, across a 12-commit branch.

(For context: this repo migrated its package manager from Bun to pnpm; the migration surface itself — bun.lock, package.json, CI workflows — never caused a conflict here, so this is pure application-code drift against a fast-moving main.)

**Comment by marcodejongh** [comment]
Rebased onto current main and retained the current Woods size scope, spray privacy, similarity caching, Kilter identity/deletion/self-alias handling, and narrow reroute reads. Independent review caught and fixed Woods hold-zero cleanup, generic Woods/spray backfill coverage, single-frame sentinel fingerprints, and database owner ordering; each relevant regression now passes, including actual local PostgreSQL temporary-table coverage.

The popular-config predicate now documents its required `bch` alias. The `(framesCount ?? 1)` fallback deliberately retains legacy null-count rows on the existing materialized-row path; only an explicit count above one selects authoritative animation frames. This does not infer that an unknown count proves animation data valid.

Validation: 945 focused tests passed across constants, Aurora/Kilter writers, backend reader/cache/privacy guards, and DB/import/backfill paths; full checks and all 53 typecheck tasks passed. One existing Kilter flush fixture requires a seeded local catalog and remains skipped. No production reads, counts, dry-run, apply, or cache actions were performed. #3985 remains open for the separately approved operator acceptance described in the runbook.

**Comment by marcodejongh** [comment]
Addressed all three minor review notes in d7ce17b4cfca53ed8178273731ddf970500e3b2d. The Aurora grammar JSDoc now describes isAuroraBoardName. The target JOIN skips materialized holds only when nonempty multi-frame text is authoritative; null/empty frames and null/single frame counts keep their original fallback.

The brittle popular-config SQL substring assertion was replaced by the full production query executed over statement-local PostgreSQL CTE fixtures. It checks corrupt sentinels/invalid IDs, valid out-of-set exclusion, future role handling, and Woods hold zero. Another real SQL test runs the resolver’s actual JOIN predicate against all five fallback boundaries. All 65 focused backend tests, full check, all 53 typecheck tasks and the mandatory hook passed. No production query, repair, cache refresh or data mutation occurred; #3985 remains open for operator acceptance.

**Comment by marcodejongh** [comment]
Addressed in a169f964a. The empty authoritative-frame regression explicitly checks that incoming frames are not projected when the persisted value is an empty string. Repair-row equality now compares the same sorted length, hold ID, frame number and state directly, avoiding JSON allocations while retaining exact equality. The existing fingerprint includes frame identity; using a hash would not add a stronger equality check.

Validation: 22 repair-helper and 68 Aurora-sync tests passed, full repository checks passed, all 53 typecheck tasks passed, and the mandatory commit hook passed. Operator acceptance remains outstanding: no production connection, count, dry-run, apply or cache refresh was performed; #3985 remains open.

**Comment by marcodejongh** [comment]
Added the explicit production status and backfill/reader explanation in 28a3fa607. The baseline is unknown, not zero: no production connection, count, dry-run, apply or cache action was performed. #3985 remains open. Full repository check, all53 typecheck tasks, body gate and commit hook passed; the unchanged typecheck retry followed a Vite+ worker stack overflow.

Two behavior details differ from the review premise:

- The target already selects nonempty multi-frame text as authoritative at [queries.ts](https://github.com/boardsesh/boardsesh/blob/28a3fa6070da81640d60728c2b34cf4ed29ffa8f/packages/backend/src/graphql/resolvers/climbs/queries.ts#L141). The JOIN now avoids transferring rows that would be discarded. Repairing stored rows does not toggle this reader policy.
- The backfill keeps NOT EXISTS and resets its UUID cursor to null at the start of every board in each invocation. A concurrent pre-cursor insert can wait for the next invocation; restarting does not resume past it. The cursor prevents repeatedly fetching a full empty-projection batch. No invocation of this script exists in the snapshot pipeline, so these details belong in the repair runbook rather than describing nonexistent snapshot automation.

The sentinel predicate is also an existing parser contract, not a guessed Aurora state name: [hold-states.ts](https://github.com/boardsesh/boardsesh/blob/28a3fa6070da81640d60728c2b34cf4ed29ffa8f/packages/board-constants/src/hold-states.ts#L403) creates `holdId=roleCode` for unknown roles, and its sentinel helper rejects `=`. Popular-config counting intentionally permits well-formed future role names; a closed known-state allowlist would lose that behavior. The actual PostgreSQL query regression covers future roles, sentinels, out-of-set holds and Woods zero.

File: packages/kilter-sync/src/sync/catalog-parse.ts [hunk 1]
```diff
@@ -176,17 +175,17 @@ export function decodeGripsClimbConcat(
...
     frameStrings.push(frameString);
   }
 
   // Aurora prefixes every frame after the first with a literal `"`.
   const frames = frameStrings.map((frameString, index) => (index === 0 ? frameString : `,"${frameString}`)).join('');
+  // Use the same first-valid projection as every other active Aurora writer.
+  // This matches the table's one-row-per-hold primary key, skips unknown-role
+  // sentinels/nonpositive IDs, and keeps the dedup fingerprint identical to
+  // the rows that are actually inserted.
+  const holds: HoldTuple[] = projectAuroraFramesToStoredRows(frames, KILTER_BOARD).rows;
```
[the remark this answers is not in the document]
Comment (marcodejongh) [reply packages/kilter-sync/src/sync/catalog-parse.ts]: Fixed in 46cb576150a0ca23730a8a45d17348ac28835019. Catalog fingerprints now hash the complete historical raw set-event stream with original frame ordinals, while board_climb_holds still receives the first-row-per-hold projection. Compatibility owner reconstruction accepts proven raw or repair-era projected stored hashes without treating projected lookup keys as dedup identity, so relights, delayed starts, and true simple owners do not fold across the lossy projection. Reroute regressions cover raw/projected stored owners and both owner orders. The preserved historical policy still does not distinguish clear-only timing or trailing empty hold frames; changing that requires a versioned fingerprint migration. No production repair or catalog sync was run.

**Comment by marcodejongh** [comment]
Thanks. I checked these against the current operator contract; no code or policy change is needed for the three non-blocking diagnostics.

- `skippedNonpositiveHoldIdTokens` is a parser diagnostic for every parsed token discarded because its hold ID is nonpositive or unsafe, including off tokens. It is printed as `skipped_nonpositive_ids`, not as candidate storage rows or deletions ([projection diagnostic](https://github.com/boardsesh/boardsesh/blob/c017a8d75a6b8a25126ceba8844662f32a5ded9b/packages/board-constants/src/hold-states.ts#L486-L493), [operator output](https://github.com/boardsesh/boardsesh/blob/c017a8d75a6b8a25126ceba8844662f32a5ded9b/packages/db/scripts/repair-board-climb-holds.ts#L329-L350)).
- `delete_rows` is the manifest's planned row work. The same output reports blockers separately, and the apply guard rejects any nonzero blocker count before the transaction can mutate rows ([guard](https://github.com/boardsesh/boardsesh/blob/c017a8d75a6b8a25126ceba8844662f32a5ded9b/packages/db/scripts/repair-board-climb-holds.ts#L308-L326)). It is intentionally useful during a blocked dry-run because it describes the work that becomes eligible after the catalog blocker is repaired.
- There is no `--expected-fingerprint-updates` option to document. Fingerprint-update membership and values are already inside the exact manifest digest, which is checked once before opening the write transaction and again after the repeatable-read rebuild under the advisory lock ([two guard calls](https://github.com/boardsesh/boardsesh/blob/c017a8d75a6b8a25126ceba8844662f32a5ded9b/packages/db/scripts/repair-board-climb-holds.ts#L628-L650)). Each eventual update also uses the reviewed old fingerprint as its compare-and-set predicate. Adding a new independent count flag would change the guard policy rather than fill a runbook omission.

The Woods hold-zero tests and backfill rerun note in `c017a8d75a6b8a25126ceba8844662f32a5ded9b` passed the focused suite and all 53 aggregate typecheck tasks. No production connection, catalog sync, cache action, or repair ran.
